### PR TITLE
Improve query optimization

### PR DESCRIPTION
**Pull request by lucasmarchd01**
This PR is a step in the process of improving our query optimization for slow-loading pages. Previously, certain pages such as source-detail and chant-list would take a significant amount of time to load due to a large number of similar or duplicate SQL queries. The solution mainly involves the use of django's `select_related` and `prefetch_related` methods that allow for improvements in the number of database hits. More specifically:
- add debug toolbar settings for development
- refactor `feast_selector_options` method to remove unnecessary nested loop
- use `select_related` and `prefetch_related` mothods to retrieve relevant data and optimize database queries by reduce the number of database hits when accessing related objects.

Here is an example of an improved query optimization on the source-detail page.

Before:
Screenshot 2023-05-30 at 11 05 06 AM copy

After:
Screenshot 2023-05-30 at 11 05 53 AM copy

### CODE REVIEW

File: django/cantusdb_project/cantusdb/settings.py [hunk 1]
```diff
@@ -61,9 +61,11 @@
     "articles",
     "django_quill",  # to provide rich-text field for articles
     "users",
+    "debug_toolbar",
```
Comment (dchiller) [inline django/cantusdb_project/cantusdb/settings.py]: Was there someway this was secure if running live? I feel like there was but I can't remember....

File: django/cantusdb_project/cantusdb/settings.py [hunk 2]
```diff
@@ -164,3 +166,9 @@
 DEFAULT_AUTO_FIELD = "django.db.models.AutoField"
 
 EMAIL_BACKEND = "django.core.mail.backends.console.EmailBackend"
+
+if DEBUG:
```
Comment (dchiller) [inline django/cantusdb_project/cantusdb/settings.py]: What is this doing?

File: django/cantusdb_project/main_app/views/chant.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 import threading
 from django.contrib import messages
 from django.contrib.postgres.search import SearchQuery, SearchRank
-from django.db.models import F, Q, QuerySet, Value
+from django.db.models import F, Q, QuerySet, Value, Prefetch
```
Comment (dchiller) [inline django/cantusdb_project/main_app/views/chant.py]: Do you use `Prefetch` in this file?

File: django/cantusdb_project/main_app/views/chant.py [hunk 4]
```diff
@@ -1463,7 +1462,7 @@ def get_feast_selector_options(source, folios):
                 folios (list of strs): A list of folios in the source.
 
             Returns:
-                zip object: A zip object combining a list of folios and Feast objects, to be unpacked in template.
+                list of tuples: A list of folios and Feast objects, to be unpacked in template.
```
Comment (dchiller) [inline django/cantusdb_project/main_app/views/chant.py]: I know this is not you, but I do not like these repeated functions!

File: django/cantusdb_project/main_app/views/source.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 from django.views.generic import DetailView, ListView, CreateView, UpdateView
-from django.db.models import Q
-from main_app.models import Source, Provenance, Century
+from django.db.models import Q, Prefetch
+from main_app.models import Source, Provenance, Century, Feast
```
Comment (dchiller) [inline django/cantusdb_project/main_app/views/source.py]: Do you use `Feast` in this file?

File: django/cantusdb_project/main_app/views/source.py [hunk 6]
```diff
@@ -42,7 +42,7 @@ def get_feast_selector_options(source, folios):
             # initialize the feast selector options with the first chant in the source that has a feast
             first_feast_chant = chants_in_source.first()
             if not first_feast_chant:
-                # if none of the chants in this source has a feast, return an empty list
+                # if none of the chants in this source has a feast, return an empty zip
```
Comment (dchiller) [inline django/cantusdb_project/main_app/views/source.py]: I think it is a list....

File: django/cantusdb_project/main_app/views/chant.py [hunk 7]
```diff
@@ -516,7 +515,7 @@ def get_queryset(self):
         search_text = self.request.GET.get("search_text")
 
         # get all chants in the specified source
-        chants = source.chant_set
+        chants = source.chant_set.select_related("feast", "office", "genre")
```
Comment (dchiller) [inline django/cantusdb_project/main_app/views/chant.py:569]: I'm confused about what this is doing...where is this additional query bit used?

**Review by dchiller — COMMENTED** [review]
I think what you are doing makes sense for optimizing the current approach. 

I think there's also a different approach -- I'm not sure that it offers much improvement in terms of load time, so maybe it is out of scope, but I think it would greatly simplify the code and the multiple nested functions. 

As I understand it, you are basically trying to come up with a list of (`folio`, `feast`) pairs....For a given source, what are all the combinations of folios and feasts (represented by chants for those feasts) that exist in the source. 

That sounds to me like a use case for a single SQL query (pseudo-coding it here):
`SELECT DISTINCT chant_tbl.folio, feast_tbl.feast_name FROM chant_tbl JOIN feast_tbl ON chant_tbl.feast = feast_tbl.id WHERE source = source_id`
I don't quite know how one would do this in django, but I wouldn't be surprised if the queryset API could do it...

File: django/cantusdb_project/cantusdb/settings.py (hunk 2 above)
Comment (lucasmarchd01) [reply django/cantusdb_project/cantusdb/settings.py]: I followed the instructions from https://django-debug-toolbar.readthedocs.io/en/latest/installation.html where it says this code is needed to configure INTERNAL_IPS correctly with Docker. Not sure exactly what's going on though. I just tested it without this and it still works so I'll probably end up removing it.

File: django/cantusdb_project/main_app/views/chant.py (hunk 3 above)
Comment (lucasmarchd01) [reply django/cantusdb_project/main_app/views/chant.py]: I was using it in a previous commit but I since changed it to select_related. Somehow those changes are not being reflected in the "files changed" tab though.

File: django/cantusdb_project/main_app/views/chant.py (hunk 4 above)
Comment (lucasmarchd01) [reply django/cantusdb_project/main_app/views/chant.py]: Yes I don't like them either... There's definitely a way this function can be refactored so that it's not nested within another function and also not repeated throughout the code.

File: django/cantusdb_project/main_app/views/chant.py (hunk 7 above)
Comment (lucasmarchd01) [reply django/cantusdb_project/main_app/views/chant.py:569]: These were necessary to include because many of the queries were caused in the chant_list.html template, specifically in:

```
</td>
<td class="text-wrap" style="text-align:center">
    {% if chant.feast %}
        <a href="{% url 'feast-detail' chant.feast.id %}" title="{{ chant.feast.description }}">{{ chant.feast.name|default:"" }}</a>
    {% endif %}
</td>
<td class="text-wrap" style="text-align:center">
    {% if chant.office %}
        <a href="{% url 'office-detail' chant.office.id %}" title="{{ chant.office.description }}">{{ chant.office.name|default:"" }}</a>
    {% endif %}
</td>
<td class="text-wrap" style="text-align:center">
    {% if chant.genre %}
        <a href="{% url 'genre-detail' chant.genre.id %}" title="{{ chant.genre.description }}">{{ chant.genre.name|default:"" }}</a>
    {% endif %}
</td>
```

File: django/cantusdb_project/cantusdb/settings.py (hunk 1 above)
Comment (lucasmarchd01) [reply django/cantusdb_project/cantusdb/settings.py]: In settings.py we have `DEBUG = bool(strtobool(os.getenv("CANTUSDB_DEBUG", "False")))` which I assume would assure it doesn't run live, but I'm not entirely sure where these environment variables are stored

Comment (ahankinson) [reply django/cantusdb_project/cantusdb/settings.py]:
```py
if DEBUG:
    INSTALLED_APPS.append("debug_toolbar")
```

File: django/cantusdb_project/cantusdb/settings.py [hunk 8]
```diff
@@ -61,9 +61,11 @@
     "articles",
     "django_quill",  # to provide rich-text field for articles
     "users",
+    "debug_toolbar",
 ]
 
 MIDDLEWARE = [
+    "debug_toolbar.middleware.DebugToolbarMiddleware",
```
Comment (ahankinson) [inline django/cantusdb_project/cantusdb/settings.py]: Same here

File: django/cantusdb_project/main_app/views/chant.py [hunk 9]
```diff
@@ -202,6 +201,58 @@ def make_suggested_chant_dict(
...
+    feast_selector_folios = []
+    # get all chants in the source, select those that have a feast
+    chants_in_source = (
+        source.chant_set.exclude(feast=None)
+        .order_by("folio", "c_sequence")
+        .select_related("feast")
+    )
+    # initialize the feast selector options with the first chant in the source that has a feast
+    first_feast_chant = chants_in_source.first()
+    if not first_feast_chant:
```
Comment (ahankinson) [inline django/cantusdb_project/main_app/views/chant.py:228]:
```py
if not first_feast_chant:
    return []
```

Then you don't need the `else` clause, and can get rid of one level of nesting.

File: django/cantusdb_project/cantusdb/settings.py (hunk 1 above)
Comment (lucasmarchd01) [reply django/cantusdb_project/cantusdb/settings.py]: Mentioning issue #467 here as I do not think this has been resolved in the production server yet.

File: django/cantusdb_project/cantusdb/settings.py (hunk 2 above)
Comment (dchiller) [reply django/cantusdb_project/cantusdb/settings.py]: Makes sense!

File: django/cantusdb_project/main_app/views/chant.py (hunk 3 above)
Comment (dchiller) [reply django/cantusdb_project/main_app/views/chant.py]: yeah, I've had that happen sometimes too...cool!

File: django/cantusdb_project/main_app/views/chant.py (hunk 7 above)
Comment (dchiller) [reply django/cantusdb_project/main_app/views/chant.py:569]: Gotcha! Nice!

File: django/cantusdb_project/cantusdb/settings.py (hunk 1 above)
Comment (jacobdgm) [reply django/cantusdb_project/cantusdb/settings.py]: Fwiw my solution has been to have a separate branch - https://github.com/jacobdgm/CantusDB/tree/toolbar - and whenever I want to test something with DjDT, I run `git checkout toolbar cantusdb/settings.py` and `git checkout toolbar cantusdb/urls.py`.

I don't think we want this turned on all the time on the staging server, where `DEBUG` should always be set to `True` I think - while it's good for testing page load times, it makes things run really slowly. But I guess it's possible to turn off all the features we don't want in the browser.

All this is to say, I have this solution that you can use if you want, but there's quite possibly a better approach, so we should discuss this tomorrow.

Comment (lucasmarchd01) [reply django/cantusdb_project/cantusdb/settings.py]: Ah okay, sounds good!

Comment (dchiller) [reply django/cantusdb_project/cantusdb/settings.py]: I'm not sure it's best to run your staging server with DEBUG = True?

Comment (jacobdgm) [reply django/cantusdb_project/cantusdb/settings.py]: I just closed #467; DEBUG is currently set to True on staging and False on Production.

I'm not sure whether to worry about having DEBUG set to True on staging. For the time being, at least, it could be quite useful to be able to check in on our query times on the ComputeCanada server (in some situations, load times are slower on CC than locally, so having the ability to test this on CC will be helpful).

A possible solution: we could start up another (limited-access) "development" server for debugging. I guess this would use up one more of our allotted IP addresses...

File: django/cantusdb_project/main_app/views/chant.py [hunk 10]
```diff
@@ -100,8 +100,7 @@ def parse_json_from_api(url: str) -> Union[list, None]:
             print(  # eventually, we should log this rather than printing it to the console
                 "Encountered an error in",
                 "ChantCreateView.get_suggested_chants.make_suggested_chant_dict",
-                "while parsing the response from",
-                f"https://cantusindex.org/json-cid/{cantus_id}:",
+                f"while parsing the response from {url}",
```
Comment (jacobdgm) [inline django/cantusdb_project/main_app/views/chant.py:103]: nice catch

File: django/cantusdb_project/main_app/views/source.py [hunk 11]
```diff
@@ -220,7 +168,13 @@ def get_queryset(self):
             )
             q_obj_filter &= indexing_search_q
 
-        return queryset.filter(q_obj_filter).distinct()
+        return (
+            queryset.filter(q_obj_filter)
+            .distinct()
```
Comment (jacobdgm) [inline django/cantusdb_project/main_app/views/source.py]: do we need `.distinct()` here? Why was this here in the first place - we don't/can't have any duplicate sources, do/can we? `.distinct`/`DISTINCT` can be a fairly heavy operation iirc, but maybe with <1000 sources it's not that bad.

"this would have been a great place for a comment," Jacob yells into the past...

**Review by jacobdgm — APPROVED** [review]
This looks good! Let's look into the one place with `.distinct` before merging, though, and either remove it or add a comment explaining why it's there...

File: django/cantusdb_project/cantusdb/settings.py (hunk 1 above)
Comment (dchiller) [reply django/cantusdb_project/cantusdb/settings.py]: Yeah, I don't think that it's an emergency in any case...just that probably the same concerns about running it in production apply to staging if those are substantially similar, which I imagine they are. 

For the longer term, something you might want to do is have the staging server no accessible to the public (eg. Cantus Ultimus staging is only accessible through the a VPN) ~~ again, I know there are bigger fish to fry right now, but maybe Katie can bring NewCantus staging into agreement with the rest of DDMAL's networking.

Comment (jacobdgm) [reply django/cantusdb_project/cantusdb/settings.py]: @kqct?

File: django/cantusdb_project/main_app/views/source.py (hunk 11 above)
Comment (lucasmarchd01) [reply django/cantusdb_project/main_app/views/source.py]: Hmm.. I tried removing distinct() and it didn't change the number of results so I'm not sure it's doing anything. It also reduces the runtime for loading the page so it might be best to remove this altogether.
